zallet/src/commands/migrate_zcashd_wallet.rs

```diff
@@ -0,0 +1,664 @@
+use std::{
+    collections::HashMap,
+    path::{Path, PathBuf},
+};
+
+use abscissa_core::Runnable;
+
+use bip0039::{English, Mnemonic};
+use secp256k1::PublicKey;
+use secrecy::SecretVec;
+use shardtree::error::ShardTreeError;
+use zaino_proto::proto::service::TxFilter;
+use zaino_state::{FetchServiceError, LightWalletIndexer};
+use zcash_client_backend::data_api::{
+    Account as _, AccountBirthday, AccountPurpose, WalletRead, WalletWrite as _, Zip32Derivation,
+    wallet::decrypt_and_store_transaction,
+};
+use zcash_client_sqlite::error::SqliteClientError;
+use zcash_keys::keys::{
+    DerivationError, UnifiedFullViewingKey,
+    zcashd::{PathParseError, ZcashdHdDerivation},
+};
+use zcash_primitives::transaction::Transaction;
+use zcash_protocol::consensus::{BlockHeight, BranchId, NetworkType, Parameters};
+use zewif_zcashd::{BDBDump, ZcashdDump, ZcashdParser, ZcashdWallet};
+use zip32::{AccountId, fingerprint::SeedFingerprint};
+
+use crate::{
+    cli::MigrateZcashdWalletCmd,
+    components::{chain_view::ChainView, database::Database, keystore::KeyStore},
+    error::{Error, ErrorKind},
+    fl,
+    prelude::*,
+    rosetta::to_chainstate,
+};
+
+use super::{AsyncRunnable, migrate_zcash_conf};
+
+/// The ZIP 32 account identifier of the zcashd account used for maintaining legacy `getnewaddress`
+/// and `z_getnewaddress` semantics after the zcashd v4.7.0 upgrade to support using
+/// mnemonic-sourced HD derivation for all addresses in the wallet.
+pub const ZCASHD_LEGACY_ACCOUNT: AccountId = AccountId::const_from_u32(0x7FFFFFFF);
+/// A source string to identify an account as being derived from the randomly generated binary HD
+/// seed used for Sapling key generation prior to the zcashd v4.7.0 upgrade.
+pub const ZCASHD_LEGACY_SOURCE: &str = "zcashd_legacy";
+/// A source string to identify an account as being derived from the mnemonic HD seed used for
+/// key derivation after the zcashd v4.7.0 upgrade.
+pub const ZCASHD_MNEMONIC_SOURCE: &str = "zcashd_mnemonic";
+
+impl AsyncRunnable for MigrateZcashdWalletCmd {
+    async fn run(&self) -> Result<(), Error> {
+        let config = APP.config();
+
+        // Start monitoring the chain.
+        let (chain_view, _chain_indexer_task_handle) = ChainView::new(&config).await?;
```

Comment:
Non-blocking: if we start any component that fires off a background task, we need to shut it down as well. This particular callsite will be altered by #237 so I will fix this then.